### PR TITLE
KTOR-973 Allow freezing of HttpClient

### PR DESCRIPTION
**Pull request by e5l**
Fixes [KTOR-973](https://youtrack.jetbrains.com/issue/KTOR-973)

### CODE REVIEW

File: ktor-client/ktor-client-core/common/src/io/ktor/client/features/HttpClientFeature.kt [hunk 1]
```diff
@@ -43,6 +43,8 @@ public fun <B : Any, F : Any> HttpClient.feature(feature: HttpClientFeature<B, F
  * @throws [IllegalStateException] if [feature] is not installed.
  */
 public operator fun <B : Any, F: Any> HttpClient.get(feature: HttpClientFeature<B, F>): F {
+    feature(feature)?.let {  return it }
```
Comment (Stexxe) [inline ktor-client/ktor-client-core/common/src/io/ktor/client/features/HttpClientFeature.kt:46]: I don't know is it idiomatic in Kotlin or not but this `.let {  return it }` hides the intent of early return.

File: ktor-client/ktor-client-core/common/src/io/ktor/client/features/HttpRedirect.kt [hunk 2]
```diff
@@ -19,30 +20,39 @@ private val ALLOWED_FOR_REDIRECT: Set<HttpMethod> = setOf(HttpMethod.Get, HttpMe
  * [HttpClient] feature that handles http redirect
  */
 public class HttpRedirect {
+    private val _checkHttpMethod = atomic(true)
```
Comment (Stexxe) [inline ktor-client/ktor-client-core/common/src/io/ktor/client/features/HttpRedirect.kt:23]: Is it in our code style to prefix private members with an underscore?

File: ktor-client/ktor-client-core/common/src/io/ktor/client/utils/SharedCollections.kt [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+
+import io.ktor.util.*
+import io.ktor.util.collections.*
+
+internal fun <K : Any, V : Any> sharedMap(): MutableMap<K, V> {
+    if (PlatformUtils.IS_NATIVE) {
+        return ConcurrentMap()
+    }
+
+    return mutableMapOf()
```
Comment (Stexxe) [inline ktor-client/ktor-client-core/common/src/io/ktor/client/utils/SharedCollections.kt:15]: Is a mutable map thread-safe and can be shared safely?

File: ktor-client/ktor-client-tests/posix/test/io/ktor/client/tests/HttpClientFreezingTest.kt [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+import io.ktor.client.tests.utils.*
+import io.ktor.utils.io.*
+import kotlin.test.*
+
+class HttpClientFreezingTest : ClientLoader() {
+
+    @Test
+    fun testRequestWithFrozenClient() = clientTests {
+        test { client ->
+            client.makeShared()
```
Comment (Stexxe) [inline ktor-client/ktor-client-tests/posix/test/io/ktor/client/tests/HttpClientFreezingTest.kt:18]: Will the clients that use `HttpClient` also call `makeShared` or is it just for testing?

File: ktor-utils/api/ktor-utils.api [hunk 5]
```diff
@@ -491,8 +491,10 @@ public final class io/ktor/util/collections/ConcurrentList : java/util/List, kot
 	public fun clear ()V
 	public fun contains (Ljava/lang/Object;)Z
 	public fun containsAll (Ljava/util/Collection;)Z
+	public fun equals (Ljava/lang/Object;)Z
```
Comment (Stexxe) [inline ktor-utils/api/ktor-utils.api:494]: Isn't it a public API change?

File: ktor-utils/common/test/io/ktor/util/PipelineContractsTest.kt [hunk 6]
```diff
@@ -166,6 +170,11 @@ class PipelineContractsTest {
 
     @Test
     fun executePipelineWithSuspension() {
+        /**
+         * Continuation doesn't support freeze in native.
+         */
+        if (PlatformUtils.IS_NATIVE) return
```
Comment (Stexxe) [inline ktor-utils/common/test/io/ktor/util/PipelineContractsTest.kt:176]: Maybe we should explicitly skip a test if the platform is native

**Review by Stexxe — CHANGES_REQUESTED** [review]
I see only scarce `PipelineContractsTest.kt` and `ConcurrentListTest`. Where are unit tests for the rest changes, e.g. for changes in `ktor-utils/common/src/io/ktor/util/pipeline/Pipeline.kt`?

File: ktor-client/ktor-client-core/common/src/io/ktor/client/features/HttpRedirect.kt (hunk 2 above)
Comment (e5l) [reply ktor-client/ktor-client-core/common/src/io/ktor/client/features/HttpRedirect.kt:23]: Yep

File: ktor-client/ktor-client-core/common/src/io/ktor/client/utils/SharedCollections.kt (hunk 3 above)
Comment (e5l) [reply ktor-client/ktor-client-core/common/src/io/ktor/client/utils/SharedCollections.kt:15]: `Shared` doesn't mean `Concurrent` here.

Comment (e5l) [reply ktor-client/ktor-client-core/common/src/io/ktor/client/utils/SharedCollections.kt:15]: This configuration shouldn't be used concurrently but can be accessed from the different threads, so it should be shared on native and can be usual on the JVM

File: ktor-client/ktor-client-tests/posix/test/io/ktor/client/tests/HttpClientFreezingTest.kt (hunk 4 above)
Comment (e5l) [reply ktor-client/ktor-client-tests/posix/test/io/ktor/client/tests/HttpClientFreezingTest.kt:18]: It's just for testing cases when user share client in the different coroutine dispatcher or create a client on the top-level scope

File: ktor-utils/api/ktor-utils.api (hunk 5 above)
Comment (e5l) [reply ktor-utils/api/ktor-utils.api:494]: Nope, the `equals` method is present by default. The difference is that we `override` it

File: ktor-utils/common/test/io/ktor/util/PipelineContractsTest.kt (hunk 6 above)
Comment (e5l) [reply ktor-utils/common/test/io/ktor/util/PipelineContractsTest.kt:176]: How should we do it? Kotlin test(and multiplatform) doesn't support the test exclusion for the platform.

File: ktor-client/ktor-client-core/common/src/io/ktor/client/features/HttpClientFeature.kt (hunk 1 above)
Comment (e5l) [reply ktor-client/ktor-client-core/common/src/io/ktor/client/features/HttpClientFeature.kt:46]: Agree, will change

File: ktor-client/ktor-client-core/common/src/io/ktor/client/utils/SharedCollections.kt (hunk 3 above)
Comment (Stexxe) [reply ktor-client/ktor-client-core/common/src/io/ktor/client/utils/SharedCollections.kt:15]: For the native, it's called`ConcurrentMap`. Maybe we should think about a more appropriate name?

File: ktor-utils/common/test/io/ktor/util/PipelineContractsTest.kt (hunk 6 above)
Comment (Stexxe) [reply ktor-utils/common/test/io/ktor/util/PipelineContractsTest.kt:176]: is there in Kotlin test something similar to https://stackoverflow.com/questions/1689242/conditionally-ignoring-tests-in-junit-4?

File: ktor-client/ktor-client-core/common/src/io/ktor/client/utils/SharedCollections.kt (hunk 3 above)
Comment (e5l) [reply ktor-client/ktor-client-core/common/src/io/ktor/client/utils/SharedCollections.kt:15]: The native map is `Concurrent`(synchronized) and  `ConcurrentMap` is shared. We have no separate `shared` map in native, so it's ok to use concurrent here.

File: ktor-utils/common/test/io/ktor/util/PipelineContractsTest.kt (hunk 6 above)
Comment (e5l) [reply ktor-utils/common/test/io/ktor/util/PipelineContractsTest.kt:176]: We have no JUnit in common, only the Kotlin Test, and as I remember it doesn't support test ignore.

Comment (e5l) [reply ktor-utils/common/test/io/ktor/util/PipelineContractsTest.kt:176]: Here is all we have in common: https://kotlinlang.org/api/latest/kotlin.test/kotlin.test/ 
So annotations only, andI'm not sure if they are working in kotlin-native

**Review by cy6erGn0m — CHANGES_REQUESTED** [review]

File: ktor-utils/common/src/io/ktor/util/collections/ConcurrentMap.kt [hunk 7]
```diff
@@ -22,20 +22,16 @@ public class ConcurrentMap<Key : Any, Value : Any>(
     private val lock: Lock = Lock(),
     initialCapacity: Int = INITIAL_CAPACITY
 ) : MutableMap<Key, Value> {
+    private var table by shared(SharedList<SharedForwardList<MapNode<Key, Value>>>(initialCapacity))
+    private var insertionOrder by shared(SharedForwardList<MapNode<Key, Value>>())
+
+    private val _size = atomic(0)
+    private val loadFactor get() = _size.value / table.size
```
[the remark this answers is not in the document]
Comment (e5l) [reply ktor-utils/common/src/io/ktor/util/collections/ConcurrentMap.kt:29]: Thanks, will fix

File: ktor-utils/common/src/io/ktor/util/collections/internal/SharedForwardList.kt [hunk 8]
```diff
@@ -6,64 +6,41 @@ package io.ktor.util.collections.internal
...
     override fun hasNext(): Boolean = current?.item != null
 
     override fun next(): T {
-        last = current
-        current = current?.next
-        skipRemoved()
-
-        return last?.item!!
+        previous = current
+        return previous?.item!!
```
[the remark this answers is not in the document]
Comment (e5l) [reply ktor-utils/common/src/io/ktor/util/collections/internal/SharedForwardList.kt:56]: +1

File: ktor-utils/common/src/io/ktor/util/collections/internal/SharedForwardList.kt [hunk 9]
```diff
@@ -6,64 +6,41 @@ package io.ktor.util.collections.internal
...
-    var current by shared(value)
-
-    init {
-        skipRemoved()
-    }
+private class ForwardListIterator<T>(head: ForwardListNode<T>) : MutableIterator<T> {
+    var previous by shared<ForwardListNode<T>?>(head)
+    val current: ForwardListNode<T>? get() = previous?.next
 
     override fun hasNext(): Boolean = current?.item != null
```
[the remark this answers is not in the document]
Comment (e5l) [reply ktor-utils/common/src/io/ktor/util/collections/internal/SharedForwardList.kt:52]: Fixed in a later commit

File: ktor-utils/common/src/io/ktor/util/pipeline/PhaseContent.kt [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+
+    @Suppress("UNCHECKED_CAST")
+    constructor(
+        phase: PipelinePhase,
+        relation: PipelinePhaseRelation
+    ) : this(phase, relation, SharedArrayList as MutableList<PipelineInterceptor<TSubject, Call>>) {
+        check(SharedArrayList.isEmpty()) { "The shared empty array list has been modified" }
+    }
+
+    var shared: Boolean by shared(true)
```
[the remark this answers is not in the document]
Comment (e5l) [reply ktor-utils/common/src/io/ktor/util/pipeline/PhaseContent.kt:25]: Yep, that's what we have

File: ktor-utils/common/src/io/ktor/util/pipeline/Pipeline.kt [hunk 11]
```diff
@@ -5,203 +5,85 @@
...
-//    }
+    @InternalAPI
+    public val isEmpty: Boolean
+        get() = interceptorsQuantity == 0
 
-    // array of phases or phase contents
-    private val phasesRaw: ArrayList<Any> = phases.mapTo(ArrayList(phases.size + 1)) {
-        it
-    }
+    private val _interceptors: AtomicRef<List<suspend PipelineContext<TSubject, TContext>.(TSubject) -> Unit>?> =
```
[the remark this answers is not in the document]
Comment (e5l) [reply ktor-utils/common/src/io/ktor/util/pipeline/Pipeline.kt:43]: `interceptors` was `Volatile` before and we don't use any atomic-specific operations, it should be the same

**Review by cy6erGn0m — APPROVED** [review]

**Review by Stexxe — APPROVED** [review]

**Review by rsinukov — APPROVED** [review]

**Pull request merged**